drahtbot/features/pr_label_acks.py

```diff
@@ -0,0 +1,182 @@
+import re
+
+from events import GitHubEvents
+from util import is_commit_hash
+
+'''
+Update PR label with ACK count (tACK, Concept ACK, etc.)
+
+@param pr: The PR object
+@param count (int): The number of ACKs
+@param type (str): The ACK type
+'''
+def update_ack_label(pr, count, type):
+    new_label = str(count) + ' ' + type
+    for label in pr.labels:
+        if label.name == new_label: # no changes
+            return
+        if re.match(r'[0-9]+ ' + type, label.name):
+            pr.remove_from_labels(label)
+            break
+
+    if count > 0:
+        pr.add_to_labels(new_label)
```

Comment:
This fails when the label doesn't exist. It might be best to ask first if people want the labels to exist. There was a previous discussion on IRC IIRC

Reply:
> This fails when the label doesn't exist

Interesting, the bot created the labels automatically for me. Maybe it's a permission issue?

> It might be best to ask first if people want the labels to exist. There was a previous discussion on IRC IIRC

Sure, should I create a GitHub issue on bitcoin/bitcoin to gain more visibility or just IRC? Also, would you have an archive link/timestamp for the previous discussion?

Reply:
> Also, would you have an archive link/timestamp for the previous discussion?

I found this discussion
https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358092;

Reply:
Thanks, interesting read. The way I see it, people didn't seem to be against it expect for @fanquake and @wumpus.

> [20:19](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358108) < wumpus> it's an interesting idea but i'm not really sure using labels for something like counting acks is a good idea, it really goes aside of the idea of labeling
> [20:20](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358109) < wumpus> there are already so many labels
> [20:21](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358110) < wumpus> ui-wise it's basically a more limited version of what bitcoinacks is doing but without the names
> [20:23](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358111) < wumpus> (and no- please don't create acked-by-wumpus acked-by-sipa etc labels xD)

> [22:04](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358117) < fanquake> I agree with wumpus, not sure about the value of “n ACKs” labels.
> [22:05](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358118) < fanquake> Also, please don’t have drahtbot start adding and removing ACK labels because some spam/unknown accounts drive by and drop ACKs on things. That’s just more noise.

I believe they were only talking about raw ACKs and not Concept/Approach/CR ACKs though. In case of raw ACKs I agree it doesn't say much about the conceptual state / review stage of a PR.

In terms of spam, I think spammers can and will find other ways to spam if they want to anyway.

Another issue I can see with labels is this: 
![image](https://user-images.githubusercontent.com/22493292/197180226-d7b156d6-3a08-457d-867c-a018b278b651.png)

There's a lot of them. So maybe instead of counting ACKs, there should only be one label per ACK type (except raw acks?).

Reply:
My idea was to have a label for an easy way to filter for pull request that have *no* current approvals, mostly as a way for reviewers to find stuff to conceptually/code review.

And for "counting" the acks, I was thinking about putting them in the summary comment (the one that also lists conflicts). However, one downside would be that the merge script would then pick up that summary comment as a review comment itself.

Reply:
I see, in this case how about adding the label `Needs Concept ACK` if there isn't any `Concept ACK` comment? And removing it once there is. This means by default a PR will start with `Needs Concept ACK` label automatically. And once the PR is Concept ACKed add labels `Needs Approach ACK` and `Needs Code Review ACK`?

Reply:
Wrong @wumpus

Reply:
> There's a lot of them. So maybe instead of counting ACKs, there should only be one label per ACK type (except raw acks?).

This is indeed noisy. I don't think that anyone will look for specific number of ACK. That said, following a "ACK(<number-of-acks>)" naming scheme seems better. If you could just sort them at the bottom of the list will probably solve this monstrosity — out of sight, out of mind. Also, "one label per ACK type" doesn't seem so useful. 

> And for "counting" the acks, I was thinking about putting them in the summary comment (the one that also lists conflicts). However, one downside would be that the merge script would then pick up that summary comment as a review comment itself.

This was also part of the aforementioned IRC discussion (messages above the linked timestamp) that stood out to me. I have to test it first, but it seems that with a minor change [in the merge script](https://github.com/bitcoin-core/bitcoin-maintainer-tools/blob/112deab9537e0bf41d74d6ac75b3bf693a26bb13/github-merge.py#L213), to check if the comment is by DrahtBot (`if (c['user']['login']] == "DrahtBot"): continue`), this could be solved. 

> I see, in this case how about adding the label `Needs Concept ACK` if there isn't any `Concept ACK` comment? And removing it once there is. This means by default a PR will start with `Needs Concept ACK` label automatically. And once the PR is Concept ACKed add labels `Needs Approach ACK` and `Needs Code Review ACK`?

This seems like a simpler, less controversial step to this labelling direction, which may help to boost interaction with PRs. Although the problem of "not all ACKs are equal" still exists, I think that this could be a nice experiment. I don't believe that there is need for multiple flags, `Needs Concept ACK` seems enough.

Note that right now there is a [Needs Conceptual Review](https://github.com/bitcoin/bitcoin/issues?q=is%3Aopen+is%3Apr+label%3A%22Needs+Conceptual+Review%22) label that is used in a different way — AFAIU is added when different opinions start to emerge regarding the change.

Reply:
Another issue I see with adding the `Needs Concept ACK` on every PR until someone comments `Concept ACK` is that it may be redundant on some PRs such as bug fixes that may have already been Concept ACKed in the corresponding issue.

There's a lot of PRs that have been merged without any Concept ACK comment: https://github.com/bitcoin/bitcoin/pulls?q=is%3Apr+sort%3Aupdated-desc+is%3Amerged+NOT+%22Concept+ACK%22+in%3Acomment

Reply:
> This was also part of the aforementioned IRC discussion (messages above the linked timestamp) that stood out to me. I have to test it first, but it seems that with a minor change [in the merge script](https://github.com/bitcoin-core/bitcoin-maintainer-tools/blob/112deab9537e0bf41d74d6ac75b3bf693a26bb13/github-merge.py#L213), to check if the comment is by DrahtBot (if (c['user']['login']] == "DrahtBot"): continue), this could be solved.

An easier way might be to just call them "reviews" instead of "ACKs", to avoid triggering the merge script.

Reply:
An idea I just had:
Drahtbot could create a summary comment when a PR is created, so it will appear at the very top and could be live-updated with the current ACKs/Approach ACKs/CR ACKs/etc. It would then be easy to glance at the state of the PR simply by reading this comment.

Reply:
>  it may be redundant on some PRs such as bug fixes that may have already been Concept ACKed in the corresponding issue.

Those PRs most of the times linked to the issue using [the supported keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword). Maybe skipping the `Needs Concept ACK` stage if this is true. This does not apply to PRs that open as follow-up (maybe a heuristic could also apply here — 'followup' on the title?) and with PRs that open with a comment on another PR as reasoning (which I am not sure that is considered Concept ACK).

> There's a lot of PRs that have been merged without any Concept ACK comment: https://github.com/bitcoin/bitcoin/pulls?q=is%3Apr+sort%3Aupdated-desc+is%3Amerged+NOT+%22Concept+ACK%22+in%3Acomment

I thought that ACK implies Concept ACK, no? That simplifies the logic. 

> An idea I just had: Drahtbot could create a summary comment when a PR is created, so it will appear at the very top and could be live-updated with the current ACKs/Approach ACKs/CR ACKs/etc. It would then be easy to glance at the state of the PR simply by reading this comment.

I don't think that this adds so much value regarding the "getting context on the state of a PR with a quick glance" use-case, because you are already in the PR page when watching this. 
That said, it's not a bad idea and I think that's what MarcoFalke said [above](https://github.com/MarcoFalke/DrahtBot/pull/1#discussion_r1001687419): "And for "counting" the acks, I was thinking about putting them in the summary comment (the one that also lists conflicts)". 

Adding to that, inspired by this part of the referenced IRC discussion
> [17:34](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358043) < jeremyrubin> when drahbot comments with the lists of PRs conflicting, a label that shows when the PR was last updated
...
[17:35](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2021-03-10#358047) < jeremyrubin> it might be convenient to show if it's something that's stale or not

I think that displaying the current status of a conflicting PR (through that summary comment) might push more people reviewing that PR to go and have a look. The implementation of this might become easier if each PR's summary contained the live-updated ACKs status (calculate once, parsed multiple times?).

Reply:
Note: The labels of a PR (such as "Needs rebase", "Needs benchmark", ... ) are already shown when hovering over the link

Reply:
> Those PRs most of the times linked to the issue using [the supported keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword). Maybe skipping the Needs Concept ACK stage if this is true. This does not apply to PRs that open as follow-up (maybe a heuristic could also apply here — 'followup' on the title?) and with PRs that open with a comment on another PR as reasoning (which I am not sure that is considered Concept ACK).

There are a lot of edge cases like this so maybe they should be addressed in a follow-up of this PR if this one ever gets merged. But I agree it might be possible to address these cases.

> I thought that ACK implies Concept ACK, no? That simplifies the logic.

Right, that could work then.

> I don't think that this adds so much value regarding the "getting context on the state of a PR with a quick glance" use-case, because you are already in the PR page when watching this.

Could be useful for large PRs with hundreds of comments. You wouldn't have to scroll all the way down, only read the first comment for a preview of the state of the PR I presume. Could also include in the summary comment whether there are unaddressed review comments for the latest commits or not.

I also agree with adding the last update, stale status, and conflicting PRs in the top summary comment.

Reply:
Also, this new top comment summary would only apply for new pulls because it needs to be at the top (for it to be useful, I think).

Reply:
Here's what the summary comment could look like. This was manually written, there's no code behind it yet.

![image](https://user-images.githubusercontent.com/22493292/198647218-dab2c166-411f-4fe2-8069-196ad80d454a.png)
I think this could be nice to have, and could also replace the current scripts so that conflicting PRs are computed the moment the PR is opened.

Reply:
Yeah, no objection. Just some nits:

* Not sure if it makes sense to differentiate reviews. Everyone uses their own description: review ACK, code ACK, ACK, utACK, tACK, tested ACK, cr ACK, ...
* Might also count NACKs?

> conflicting PRs are computed the moment the PR is opened.

It takes a few minutes to calculate them, so I think this is unrelated in any case.

Reply:
> Not sure if it makes sense to differentiate reviews. Everyone uses their own description: review ACK, code ACK, ACK, utACK, tACK, tested ACK, cr ACK, ...
    Might also count NACKs?

Yes, I plan to include all types of ACKs, this was just an example.
Also, I currently parse ACKs with aliases such as: `Code Review ACK`, `CR ACK`, `crACK`, `re-crACK` etc will all be included in the `Code Review ACK` cell, same for `tACK`, `utACK`, `ACK`, `reACK`, `re-ACK`.

Reply:
Ok, maybe ask other devs if anyone is objecting/liking this, since it requires a summary comment on *all* pulls.

Reply:
Looks good, I agree with both of MarcoFalke's points. I think to differentiate between Concept ACKs and the rest is enough. If you go this way, you could just have "Concept ACKs",  "Reviews", "Stale Reviews" ([as also mentioned above](https://github.com/MarcoFalke/DrahtBot/pull/1#discussion_r1005508053)) and "NACKs".

I assume clicking on a reviewer's name takes you to their review?

Reply:
> Looks good, I agree with both of MarcoFalke's points. I think to differentiate between Concept ACKs and the rest is enough. If you go this way, you could just have "Concept ACKs", "Reviews", "Stale Reviews" (https://github.com/MarcoFalke/DrahtBot/pull/1#discussion_r1005508053) and "NACKs".

Interesting, I would actually prefer to list different types of ACKs, which are: (N)ACKs, Concept (N)ACKs, Approach (N)ACKs, Code Review ACKs, NACKs, and Stale ACKs. I would omit them from the table if the count is 0 though.

It gives a bit more details on the review stage of the PR IMO, and whether it is ready for merge soon or not.

<details>
<summary>Examples</summary>


| ACKs | Count | Reviewers |
| --- | ----- | --------- |
| **NACKs**  | 1 | [aureleoules](https://github.com/aureleoules) |
| **ACKs** | 3 | [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke) [kouloumos](https://github.com/MarcoFalke) |
| Concept ACKs | 2 | [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke) |
| Approach ACKs  | 1 | [aureleoules](https://github.com/aureleoules) |
| Code Review ACKs  | 1 | [Satoshi](https://github.com/satoshi) |
| _Stale_ ACKs | 1 | [Satoshi](https://github.com/satoshi)

vs

| Reviews | Count | Reviewers |
| --- | ----- | --------- |
| **NACKs**  | 1 | [aureleoules](https://github.com/aureleoules) |
| Concept ACK | 3 | [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke) [kouloumos](https://github.com/MarcoFalke) |
| Reviews | 12 | [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke) [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke) [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke) [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke) [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke) [aureleoules](https://github.com/aureleoules) [MarcoFalke](https://github.com/MarcoFalke)|
| _Stale_ ACKs | 1 | [Satoshi](https://github.com/satoshi)

</details>


> I assume clicking on a reviewer's name takes you to their review?

Yes!

> Is it too much to grasp the opportunity of mentioning code review terms to also drop somewhere in there a hyperlink to [bitcoin/bitcoin@master/CONTRIBUTING.md#decision-making-process](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md?rgh-link-date=2022-10-28T15%3A24%3A28Z#decision-making-process)?

Yes, could be nice to have it in the summary comment.

Reply:
@MarcoFalke I see that you're rewriting parts of the bot in Rust. Would you like me to rewrite this PR in Rust? I'm not familiar with it but would love the learning exercise!

Reply:
Sure, why not, if you want to